src/util/std_types.h

```diff
@@ -768,6 +768,15 @@ class array_typet:public type_with_subtypet
     add(ID_size, std::move(_size));
   }
 
+  /// The type of the index expressions into any instance of this type.
+  typet index_type() const;
```

Comment:
I am concerned that this may be confusing: we could end up in a situation where `array.index().type() != array.index_type()`!

Reply:
I'd like to enforce that, going forward, using a precondition.